### ISSUE TITLE
Filter out the following reviews in the optimizer 

### ISSUE THREAD

**Issue by user1823** [open-spaced-repetition/fsrs4anki#270]
**Which module is related to your feature request?**
Optimizer

**Is your feature request related to a problem? Please describe.**
When the optimizer is fed on data containing incomplete reviews/ outliers, it tends to produce incorrect parameters.

**Describe the solution you'd like**
Before optimization begins, the optimizer should filter out the following cards/reviews:
- In cards having a manual reset (both having and not having a manual rating, see https://github.com/open-spaced-repetition/fsrs4anki-helper/commit/b2090043d6e073eb69b0206e7419e73dc48c57e4), the **ratings before the manual reset** should be filtered out. However, the ratings after the manual reset should be preserved and used for optimization.
- **All the ratings of suspended cards** should be filtered out (because they are more likely to be outliers).
- **All the ratings of cards in which the first review is not of learn type** should be filtered out. To make it clear, I want the optimizer to filter out all the ratings of even those cards in which the first review is of relearn type.

You implemented something of this sort in https://github.com/open-spaced-repetition/fsrs4anki/commit/ff773098dfd11dd09f8a9b4068fa2f17d0f4c2ca. But, I am not sure if it is exactly the same as I want it to be of not.

**Comment by L-M-Sherlock** [comment]
> * **All the ratings of suspended cards** should be filtered out (because they are more likely to be outliers).

They are hardly filtered out. Because suspending is not recorded in the revlog.

**Comment by L-M-Sherlock** [comment]
> * In cards having a manual reset (both having and not having a manual rating, see [open-spaced-repetition/fsrs4anki-helper@b209004](https://github.com/open-spaced-repetition/fsrs4anki-helper/commit/b2090043d6e073eb69b0206e7419e73dc48c57e4)), the **ratings before the manual reset** should be filtered out. However, the ratings after the manual reset should be preserved and used for optimization.

The current behavior is keeping  the **ratings before the manual reset**:  https://github.com/open-spaced-repetition/fsrs4anki/commit/ff773098dfd11dd09f8a9b4068fa2f17d0f4c2ca

**Comment by L-M-Sherlock** [comment]
> * **All the ratings of cards in which the first review is not of learn type** should be filtered out. To make it clear, I want the optimizer to filter out all the ratings of even those cards in which the first review is of relearn type.

It has been implemented in https://github.com/open-spaced-repetition/fsrs4anki/commit/ff773098dfd11dd09f8a9b4068fa2f17d0f4c2ca.

**Comment by user1823** [comment]
> > * **All the ratings of suspended cards** should be filtered out (because they are more likely to be outliers).
> 
> They are hardly filtered out. Because suspending is not recorded in the revlog.

The collection file definitely contains the information about which cards are suspended. So, there should be some way to identify and filter out these cards.

> The current behavior is keeping the **ratings before the manual reset**: [ff77309](https://github.com/open-spaced-repetition/fsrs4anki/commit/ff773098dfd11dd09f8a9b4068fa2f17d0f4c2ca)

So, we want to change this behaviour.

**Comment by L-M-Sherlock** [comment]
> The collection file definitely contains the information about which cards are suspended. So, there should be some way to identify and filter out these cards.

No, the collection file only contains the information that which cards are suspended instead of which cards have been suspended and unsuspended later.

**Comment by user1823** [comment]
> No, the collection file only contains the information that which cards are suspended instead of which cards have been suspended and unsuspended later.

I meant only those cards which are currently suspended (not those which were unsuspended later).

**Comment by L-M-Sherlock** [comment]
> Due to those reviews, the actual retention is higher than the predicted retention in the fixed forgetting curve version in low retention areas. Maybe you have noted that in the calibration graph of parametrized curve version, very few reviews are predicted with low retention. So its blue line is aligned to the orange line well.

I think we should filter out those super-overdue reviews (they may be suspended for a long time and unsuspended for some reason). These reviews are biased or heterogeneous. They could mess up the algorithm.

_Originally posted by @L-M-Sherlock in https://github.com/open-spaced-repetition/fsrs4anki/issues/239#issuecomment-1565973059_

In my view, those cards unsuspended later will induce the problem I described above.

**Comment by L-M-Sherlock** [comment]
For example, a card is suspended following the initial learning and then unsuspended after a hundred days, upon which it is reviewed. This review will be affected by many factors.

**Comment by user1823** [comment]
I think that a better way to describe such cards is **super-overdue cards** rather than unsuspended cards. This is because
- cards can be unsuspended on the same day they were suspended (e.g. suspended by mistake)
- cards can be unsuspended after a few days but still before they were next due (so suspending the card had no effect)
- cards can be super-overdue even if they were never suspended (e.g. cards in a separate deck that wasn't reviewed since long)

But, I don't know how much overdue should a card be that the optimizer should filter it out.

Also, my rationale for filtering out the cards that are currently suspended:
- they can be related to something that the user is not currently studying (so the forgetting rate of that material may not be similar to what the user is currently studying)
- they can be poorly formulated cards / leeches (so the forgetting rate of those items is much higher than that of the other items).

**Comment by Expertium** [comment]
Is there a way to filter cards (in Anki itself, not in the optimizer) that have been manually reset?

**Comment by L-M-Sherlock** [comment]
> Is there a way to filter cards (in Anki itself, not in the optimizer) that have been manually reset?

Could you share your latest collection file with me? I want to find a way to filter them.

**Comment by Expertium** [comment]
https://drive.google.com/file/d/1G5qLsHxoIz0n_JgBcVfZr26ZnG_S_Swt/view?usp=sharing

**Comment by user1823** [comment]
Taking hint from a recent patch to the helper add-on (https://github.com/open-spaced-repetition/fsrs4anki-helper/commit/74f17ab478752862e44bfe2ba0c43fbcc425aac1), I would want you to filter out from the optimizer the reviews that are done in filtered decks with rescheduling disabled (not to be confused with FSRS helper rescheduling).

**Comment by L-M-Sherlock** [comment]
> would want you to filter out from the optimizer the reviews that are done in filtered decks with rescheduling disabled 

The current optimizer has filtered out this reviews.

**Comment by L-M-Sherlock** [comment]
> This probably suggests that Anki doesn't clear the revlogs of cards that have been deleted.

That would be a **major** problem for optimization, if that's true. I hope Sherlock can figure out how so solve it.

_Originally posted by @Expertium in https://github.com/open-spaced-repetition/fsrs4anki/issues/314#issuecomment-1611709296_

Maybe we should also filter out those review logs of deleted cards?

**Comment by L-M-Sherlock** [comment]
https://colab.research.google.com/github/open-spaced-repetition/fsrs4anki/blob/v3.25.7/fsrs4anki_optimizer.ipynb

This version filters out the deleted cards' revlogs.

**Comment by L-M-Sherlock** [comment]
> * **All the ratings of suspended cards** should be filtered out (because they are more likely to be outliers).

They are not outliers in my case. I archived some cards via `suspend` in my collection. The review logs of these cards are useful.

**Comment by user1823** [comment]
> > * **All the ratings of suspended cards** should be filtered out (because they are more likely to be outliers).
>
> They are not outliers in my case. I archived some cards via `suspend` in my collection. The review logs of these cards are useful.

What do you exactly mean by "archiving"? If the material is similar to what you are currently learning, why are those cards suspended?

Also, what is your opinion about the following argument?
> My rationale for filtering out the cards that are currently suspended:
> 
> * they can be related to something that the user is not currently studying (so the forgetting rate of that material may not be similar to what the user is currently studying)
> * they can be poorly formulated cards / leeches (so the forgetting rate of those items is much higher than that of the other items).

In addition, what about filtering the ratings before a manual reset?

> > * In cards having a manual reset (both having and not having a manual rating, see [open-spaced-repetition/fsrs4anki-helper@b209004](https://github.com/open-spaced-repetition/fsrs4anki-helper/commit/b2090043d6e073eb69b0206e7419e73dc48c57e4)), the **ratings before the manual reset** should be filtered out. However, the ratings after the manual reset should be preserved and used for optimization.
> 
> The current behavior is keeping the **ratings before the manual reset**.

**Comment by L-M-Sherlock** [comment]
> What do you exactly mean by "archiving"? If the material is similar to what you are currently learning, why are those cards suspended?

They are not similar to my current learning. But they were all written by myself. In the exemplary collection (my collection optimized in the notebook), the most cards were suspended.

**Comment by L-M-Sherlock** [comment]
> In addition, what about filtering the ratings before a manual reset?

The first rating after a manual reset is ambiguous. The rating of manual reset is `0`.

**Comment by L-M-Sherlock** [comment]
> * they can be poorly formulated cards / leeches (so the forgetting rate of those items is much higher than that of the other items).

Why not just filter out those cards whose number of lapses exceeds 8?

**Comment by user1823** [comment]
> > In addition, what about filtering the ratings before a manual reset?
> 
> The first rating after a manual reset is ambiguous. The rating of manual reset is `0`.

Treat the first rating after the manual rating as the first rating. After manual reset, Anki treats the cards as **new**.

By the way, this problem is more important when you consider the ratings before the manual reset because the rating of manual reset is `0` and then the optimizer can't make any prediction of retention for the next review.

When we consider the card to be new after a manual reset, the next rating is treated as the first rating and no such problem arises.

> > * they can be poorly formulated cards / leeches (so the forgetting rate of those items is much higher than that of the other items).
> 
> Why not just filter out those cards whose number of lapses exceeds 8?

Now, I think that we should leave it to the user whether they want to train the parameters on the suspended cards or not. They can easily control this by exporting either all cards or only unsuspended cards from the Anki Browser.

**Comment by L-M-Sherlock** [comment]
> Treat the first rating after the manual reset as the first rating. After manual reset, Anki treats the cards as **new**.

I think it is not simple. Here is a example:

![image](https://user-images.githubusercontent.com/5121875/232280498-4452ab5a-ef5b-450f-89dc-3d6dcb968d77.png)

Source: https://github.com/open-spaced-repetition/fsrs4anki-helper/issues/64

The manual reset has its own interval.

> Now, I think that we should leave it to the user whether they want to train the parameters on the suspended cards or not. They can easily control this by exporting either all cards or only unsuspended cards from the Anki Browser.

By the way, some users may change the default action of leeches (i.e. only tagged them).

**Comment by user1823** [comment]
> > Treat the first rating after the manual reset as the first rating. After manual reset, Anki treats the cards as **new**.
> 
> I think it is not simple. Here is a example:
> 
> ![image](https://user-images.githubusercontent.com/5121875/232280498-4452ab5a-ef5b-450f-89dc-3d6dcb968d77.png)
> 
> Source: [open-spaced-repetition/fsrs4anki-helper#64](https://github.com/open-spaced-repetition/fsrs4anki-helper/issues/64)
> 
> The manual reset has its own interval.

This is not "Reset". This is "Set Due Date".

Sorry for the confusion, but I was talking about the "Forget" feature in Anki.

But, there can be two kinds of manual reset in the revlogs:
- One kind of manual reset has a manual rating (0) and is produced by recent versions of Anki.
- The other kind of manual reset doesn't have any rating in the revlog.

You developed the code to find them in the Helper: https://github.com/open-spaced-repetition/fsrs4anki-helper/commit/b2090043d6e073eb69b0206e7419e73dc48c57e4)

> By the way, some users may change the default action of leeches (i.e. only tagged them).

If you think that filtering out cards with more than 8 lapses would be beneficial for all users, go for it. But, I don't have any strong opinions about it.

**Comment by user1823** [comment]
We might not be filtering out the suspended cards in the training data. But, when we will later use a separate dataset for evaluation, we should **filter out the suspended cards in the evaluation data**. The reason being that even if the revlogs of the suspended cards can be helpful in training the parameters, those revlogs are of no value in the evaluation process (because those cards are not being reviewed).

**Comment by L-M-Sherlock** [comment]
Record some useful information:

The code used to log `Forget` & `Set due date`:

https://github.com/ankitects/anki/blob/14bc02b4316bb693f3e166e20d76b767e139550d/rslib/src/revlog/mod.rs#L87-L108

The code of `Forget`:

https://github.com/ankitects/anki/blob/14bc02b4316bb693f3e166e20d76b767e139550d/rslib/src/scheduler/new.rs#L151-L190

The code of `Set due date`:

https://github.com/ankitects/anki/blob/14bc02b4316bb693f3e166e20d76b767e139550d/rslib/src/scheduler/reviews.rs#L100-L146

The difference of revlogs between `Forget` and `Set due date`:

The `ivl` and `factor` of `Forget` revlog are zero. For `Set due date` revlog, they are non-zero.

**Comment by L-M-Sherlock** [comment]
@user1823, should we filter out the revlogs of `Set due date`? I skip these reviews in the helper add-on.

**Comment by user1823** [comment]
> @user1823, should we filter out the revlogs of `Set due date`? I skip these reviews in the helper add-on.

If we assume that the user didn't recall the card before using `set due date`, we should filter out the revlogs of `set due date` (but not the revlogs before or after the `set due date` rating).

But, the things get complicated if we consider that the user might have recalled the card first and then used `set due date`. 

However, I think that the second case is unlikely/rare. So, it is better to filter out the revlogs of `set due date` (but not the revlogs before or after the `set due date` rating).

> The difference of revlogs between Forget and Set due date:
> The ivl and factor of Forget revlog are zero. For Set due date revlog, they are non-zero.

As I said before, there is one more type of Forget, which was produced by older versions of Anki. In this type, there is no Manual rating in the revlogs.

Example:
![image](https://user-images.githubusercontent.com/38519777/225714554-ed1ced5c-eeab-487f-988d-6b9e44969ef8.png)
Source: https://github.com/open-spaced-repetition/fsrs4anki-helper/issues/49

You developed code to find such cases in the helper add-on:
https://github.com/open-spaced-repetition/fsrs4anki-helper/blob/34af286719944425abdc3c0538a5eedeb7e36e52/reschedule.py#L201-L219

**Comment by L-M-Sherlock** [comment]
OK. My current plan is 

1. skip all `Set due date` reviews
2. filter out all reviews which are previous to the last `Learn` review for each card

**Comment by user1823** [comment]
> OK. My current plan is
> 
> 1. skip all `Set due date` reviews
> 2. filter out all reviews which are previous to the last `Learn` review for each card

It looks good, except that it should be the first `Learn` review in the last set of `Learn` reviews.

The reason being that there can be multiple learn reviews in the same learning stage when there are multiple learning steps.

**Comment by user1823** [comment]
@L-M-Sherlock, one small suggestion:

Filtering of reviews before the first `Learn` review in the last learning stage should happen before the cards whose first review is not of learn type are filtered (https://github.com/open-spaced-repetition/fsrs4anki/commit/ff773098dfd11dd09f8a9b4068fa2f17d0f4c2ca).

This will ensure that we have an equivalent of https://github.com/open-spaced-repetition/fsrs4anki-helper/commit/b2090043d6e073eb69b0206e7419e73dc48c57e4.

**Comment by user1823** [comment]
> https://github.com/open-spaced-repetition/fsrs4anki/blob/Fix/consider-manual-schedule-revlog/fsrs4anki_optimizer.ipynb
> 
> Could you help me test it?

I used the following link:
https://colab.research.google.com/github/open-spaced-repetition/fsrs4anki/blob/Fix/consider-manual-schedule-revlog/fsrs4anki_optimizer.ipynb

It failed with the following error:
<img width="693" alt="image" src="https://github.com/open-spaced-repetition/fsrs4anki/assets/92206575/55b62ae7-1104-40d8-886e-8379d7393a14">

By the way, testing whether the optimizer filters the correct reviews is not possible. The only thing I could test is whether it runs smoothly or it leads to some errors.

**Comment by L-M-Sherlock** [comment]
My fault. I forget to release it. Now you can test it in here: https://colab.research.google.com/github/open-spaced-repetition/fsrs4anki/blob/v3.25.8/fsrs4anki_optimizer.ipynb

**Comment by user1823** [comment]
So, is it executed after the new code in https://github.com/open-spaced-repetition/fsrs4anki/pull/325 is executed or before the new code is executed?

**Comment by L-M-Sherlock** [comment]
After:

https://github.com/open-spaced-repetition/fsrs4anki/blob/aab50e17c84b6fb82d425879055c1934010aa935/package/fsrs4anki_optimizer/fsrs4anki_optimizer.py#L405-L408
